Cesium3DTilesSelection/test/TestCreditSystem.cpp

```diff
@@ -198,3 +199,285 @@ TEST_CASE("Test setting showOnScreen on credits") {
   CHECK(creditSystem.shouldBeShownOnScreen(credit1) == true);
   CHECK(creditSystem.shouldBeShownOnScreen(credit2) == true);
 }
+
+TEST_CASE("Test CreditSystem with CreditSources") {
+  CreditSystem creditSystem;
+  CreditSource sourceA(creditSystem);
+  CreditSource sourceB(creditSystem);
+
+  std::string html0 = "<html>Credit0</html>";
+  std::string html1 = "<html>Credit1</html>";
+  std::string html2 = "<html>Credit2</html>";
+
+  SUBCASE("can create credits from multiple sources") {
+    Credit credit0 = creditSystem.createCredit(sourceA, html0);
+    Credit credit1 = creditSystem.createCredit(sourceB, html1);
+
+    CHECK(creditSystem.getCreditSource(credit0) == &sourceA);
+    CHECK(creditSystem.getHtml(credit0) == html0);
+    CHECK(creditSystem.getCreditSource(credit1) == &sourceB);
+    CHECK(creditSystem.getHtml(credit1) == html1);
+  }
+
+  SUBCASE("credits become invalid when their source is destroyed") {
+    std::optional<Credit> credit;
+
+    {
+      CreditSource tempSourceA(creditSystem);
+      credit = creditSystem.createCredit(tempSourceA, html0);
+      CHECK(creditSystem.getCreditSource(*credit) == &tempSourceA);
+    }
+
+    // tempSourceA is destroyed here.
+    // The credit should no longer have a source.
+    CHECK(creditSystem.getCreditSource(*credit) == nullptr);
+    // Getting HTML from a credit with no source should not crash and it provide
+    // a non-empty (error) message.
+    CHECK(!creditSystem.getHtml(*credit).empty());
+
+    // Creating a new credit from a different source should still work.
+    CreditSource tempSourceB(creditSystem);
+    Credit credit1 = creditSystem.createCredit(tempSourceB, html1);
+    CHECK(creditSystem.getCreditSource(credit1) == &tempSourceB);
+    CHECK(creditSystem.getHtml(credit1) == html1);
+  }
+
+  SUBCASE("when the source of a credit last frame is destroyed, that credit is "
+          "not reported") {
+    std::unique_ptr<CreditSource> pTempSourceA =
+        std::make_unique<CreditSource>(creditSystem);
+    Credit credit0 = creditSystem.createCredit(*pTempSourceA, html0);
+
+    creditSystem.addCreditReference(credit0);
+    const CreditsSnapshot& snapshot0 = creditSystem.getSnapshot();
+    CHECK(snapshot0.currentCredits.size() == 1);
+    CHECK(snapshot0.currentCredits[0] == credit0);
+
+    // Remove the credit reference, which will add it to the list of "no longer
```

Comment:
`_creditsToNoLongerShowThisSnapshot` got removed, so this comment is outdated, right?